### PR TITLE
Replace deprecated wait.poll

### PR DESCRIPTION
**Pull request by Affan-7**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
This PR replaces the deprecated `wait.Poll()` function with the `wait.PollUntilContextTimeout()` function.

**Which issue(s) this PR fixes**:
Related to #3835
1 out of 4 tasks

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

**Comment by carlory** [comment]
/lgtm

/assign @rainbowmango 

for workflow and approval.

**Review by RainbowMango — APPROVED** [review]
/approve
Thanks~

**Comment by Affan-7** [comment]
The tests were failing, here are the logs: https://github.com/karmada-io/karmada/actions/runs/5783869909/job/15676511801

Should we increase the default value of `j.Wait`? Or there is some other problem?

**Comment by RainbowMango** [comment]
It might be flak, let's give it another try.

**Comment by RainbowMango** [comment]
> Should we increase the default value of j.Wait? Or there is some other problem?

I'm not sure yet. let's give it another try and try to find the root cause.

I'm curious that the test failed in `42` seconds, but the default waiting time(`j.Wait`) is `60` seconds.

Just echo the log here: 
https://github.com/karmada-io/karmada/actions/runs/5787607401/job/15697196471?pr=3906#step:5:2723
```
[FAILED] [42.440 seconds]
[cluster joined] reschedule testing Deployment propagation testing [AfterEach] testing clusterAffinity of the policy when the ReplicaScheduling of the policy is nil, reschedule testing [needCreateCluster]
  [AfterEach] /home/runner/work/karmada/karmada/test/e2e/rescheduling_test.go:176
  [It] /home/runner/work/karmada/karmada/test/e2e/rescheduling_test.go:271

  Captured StdOut/StdErr Output >>
  I0808 02:23:49.855080   46074 deployment.go:199] The ResourceBinding(karmadatest-2dj87/deploy-tjnp2-deployment) schedule result is: member1,member2
  cluster(member-e2e-lhr) is joined successfully
  I0808 02:23:59.098817   46074 deployment.go:199] The ResourceBinding(karmadatest-2dj87/deploy-tjnp2-deployment) schedule result is: member-e2e-lhr,member1,member2
  I0808 02:23:59.109167   46074 deployment.go:199] The ResourceBinding(karmadatest-2dj87/deploy-tjnp2-deployment) schedule result is: member-e2e-lhr,member1,member2
  E0808 02:23:59.135302   46074 unjoin.go:239] Failed to delete cluster object. cluster name: member-e2e-lhr, error: context deadline exceeded
  E0808 02:23:59.136891   46074 unjoin.go:174] Failed to delete cluster object. cluster name: member-e2e-lhr, error: context deadline exceeded
```

File: pkg/controllers/certificate/cert_rotation_controller.go [hunk 1]
```diff
@@ -165,7 +165,7 @@ func (c *CertRotationController) syncCertRotation(secret *corev1.Secret) error {
 
 	var newCertData []byte
 	klog.V(1).Infof("Waiting for the client certificate to be issued")
-	err = wait.Poll(1*time.Second, 5*time.Minute, func() (done bool, err error) {
+	err = wait.PollUntilContextTimeout(context.TODO(), 1*time.Second, 5*time.Minute, false, func(ctx context.Context) (done bool, err error) {
```
Comment (Jorropo) [inline pkg/controllers/certificate/cert_rotation_controller.go:168]: Theses are not equivalent, I think you meant to use `PollWithContextTimeout` as the docs said to do, however this does not exists.

Instead you can do:
```suggestion
  ctx, cancel := context.WithTimeout(context.TODO(), 5*time.Minute)
  defer cancel()
	err = wait.PollWithContext(ctx, 1*time.Second, 5*time.Minute, func(ctx context.Context) (done bool, err error) {
```

File: pkg/controllers/certificate/cert_rotation_controller.go [hunk 2]
```diff
@@ -165,7 +165,7 @@ func (c *CertRotationController) syncCertRotation(secret *corev1.Secret) error {
...
 		csr, err := c.KubeClient.CertificatesV1().CertificateSigningRequests().Get(context.TODO(), csr, metav1.GetOptions{})
```
Comment (Jorropo) [inline pkg/controllers/certificate/cert_rotation_controller.go:169]:
```suggestion
		csr, err := c.KubeClient.CertificatesV1().CertificateSigningRequests().Get(ctx, csr, metav1.GetOptions{})
```
Even if it's not passed up, this can be passed down.

**Review by Jorropo — CHANGES_REQUESTED** [review]
If you change all of the `PollUntilContextTimeout` like I showed I belive it would work. :slightly_smiling_face:

File: pkg/controllers/certificate/cert_rotation_controller.go (hunk 1 above)
Comment (Affan-7) [reply pkg/controllers/certificate/cert_rotation_controller.go:168]: Thanks for your collaboration @Jorropo, but unfortunately the `PollWithContext` method is also deprecated.

Reference: https://pkg.go.dev/k8s.io/apimachinery@v0.27.3/pkg/util/wait#PollWithContext

Comment (Jorropo) [reply pkg/controllers/certificate/cert_rotation_controller.go:168]: Ah right mb.

**Comment by RainbowMango** [comment]
/remove-approve
We need to figure out the root cause, probably affected by https://github.com/kubernetes/kubernetes/issues/119533

**Comment by Affan-7** [comment]
## Here is what I did:

**TL:DR;**

Value of `j.Wait` was zero in the both cases `wait.Poll` and `wait.PollUntilContextTimeout()`. Since the `wait.Poll()` is deprecated it no longer returns any errors. That's why `wait.Poll()` was passing the test and `wait.PollUntilContextTimeout()` was failing. I specified the value of `j.Wait` explicitly under the test file and everything works fine as expected. 

---

I added `fmt.Printf("+++++ value of j.Wait: %v\n", j.Wait)` before `wait.PollUntilContextTimeout(context.TODO(), 1*time.Second, j.Wait, .....)`. And I ran the test on the my fork of Karmada, using github actions. I was surprised that the value of `j.Wait` was `zero`.

Reference: https://github.com/Affan-7/karmada/actions/runs/5841035785/job/15840850551#step:5:2798

So I run the test again with the deprecated `wait.Poll()` method. But the test passed! Then I ran it again with `fmt.Printf("+++++ value of j.Wait: %v\n", j.Wait)` and the value of `j.Wait` was also `zero`.

Reference: https://github.com/Affan-7/karmada/actions/runs/5841648595/job/15842010633#step:5:2999

The value of `j.Wait` was still `zero` but the test case weren't failing. Because the `wait.Poll()` method is deprecated and doesn't return errors anymore. I found this in the documentation of `wait.Poll`.

``` 
Deprecated: This method does not return errors from context, use PollWithContextTimeout. Note that the new method will no longer return ErrWaitTimeout and instead return errors defined by the context package. Will be removed in a future release.
```

Reference: https://pkg.go.dev/k8s.io/apimachinery/pkg/util/wait#Poll

I was assuming that the default value of `j.Wait` should be `60` because of `flags.DurationVar(&j.Wait, "wait", 60*time.Second, ....)` under the `unjoin.go`. But this block of code doesn't get executed during the test (because it's for CLI). Therefore the value of 'j.Wait' remains `zero`.

So I added the value of `j.Wait` explicitly under `rescheduling_test.go` by using `Wait: 60 * time.Second,`. It's very similar to the `Wait: 5 * options.DefaultKarmadactlCommandDuration,` on the [line 441 of the `karmadactl_test.go`](https://github.com/karmada-io/karmada/blob/6dea12190e9752eef937c2461e0223a5fe88ee6a/test/e2e/karmadactl_test.go#L441C15-L441C15).

After specifying the value of `j.Wait` under the `rescheduling_test.go` all the tests were passing successfully.
Reference: https://github.com/Affan-7/karmada/actions/runs/5841642875/job/15841998395#step:5:3461

**Comment by Affan-7** [comment]
Can you please merge this @RainbowMango :blush:. Or is there  something that should I improve?

**Comment by RainbowMango** [comment]
Thanks @Affan-7 for the excellent analysis, I think it is acceptable to set a timeout [here](https://github.com/karmada-io/karmada/pull/3906/files#diff-b6331f265a97596e2848f6e505111c716f47659dd9057127740cbf1ee1d0d484R124-R184).

I wonder if this is affected by https://github.com/kubernetes/kubernetes/issues/119533? 

I'm hesitant to do it before https://github.com/kubernetes/kubernetes/issues/119533.

File: pkg/karmadactl/register/register.go [hunk 3]
```diff
@@ -509,7 +509,7 @@ func (o *CommandRegisterOption) constructKarmadaAgentConfig(bootstrapClient *kub
 	}
 
 	klog.V(1).Infof("Waiting for the client certificate to be issued")
-	err = wait.Poll(1*time.Second, o.Timeout, func() (done bool, err error) {
+	err = wait.PollUntilContextTimeout(context.TODO(), 1*time.Second, o.Timeout, false, func(ctx context.Context) (done bool, err error) {
```
Comment (zhzhuang-zju) [inline pkg/karmadactl/register/register.go:512]: Why `immediate` here is set to `false`? `condition` can be invoked immediately.

File: pkg/karmadactl/unjoin/unjoin.go [hunk 4]
```diff
@@ -223,7 +223,7 @@ func (j *CommandUnjoinOption) deleteClusterObject(controlPlaneKarmadaClient *kar
 	}
 
 	// make sure the given cluster object has been deleted
-	err = wait.Poll(1*time.Second, j.Wait, func() (done bool, err error) {
+	err = wait.PollUntilContextTimeout(context.TODO(), 1*time.Second, j.Wait, false, func(ctx context.Context) (done bool, err error) {
```
Comment (zhzhuang-zju) [inline pkg/karmadactl/unjoin/unjoin.go:226]: ditto~

File: pkg/karmadactl/register/register.go (hunk 3 above)
Comment (RainbowMango) [reply pkg/karmadactl/register/register.go:548]: That is because this behavior is equipment with previous.
`wait.Poll` always waits the interval before the run of `condition`.

File: pkg/karmadactl/unjoin/unjoin.go (hunk 4 above)
Comment (RainbowMango) [reply pkg/karmadactl/unjoin/unjoin.go:242]: ditto~
answered at the last comment.

**Comment by RainbowMango** [comment]
Just rebased this and fixed the golint issues.
/assign @XiShanYongYe-Chang

File: pkg/karmadactl/register/register.go (hunk 3 above)
Comment (RainbowMango) [reply pkg/karmadactl/register/register.go]: That's the idea not run the condition so rush.

**Review by XiShanYongYe-Chang — COMMENTED** [review]
Thanks~
/lgtm
